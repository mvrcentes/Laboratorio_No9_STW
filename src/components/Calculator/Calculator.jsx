import React, { useState } from "react"
import { Button } from "../Button/Button"

import style from "./Calculator.module.css"

export const Calculator = ({ children }) => {
    const numbers = [7, 8, 9, 4, 5, 6, 1, 2, 3, ".", 0, "←"]
    const [inputValue, setInputValue] = useState("0")
    const [previousValue, setPreviousValue] = useState(null)
    const [operator, setOperator] = useState(null)

    const handleDisplay = (value) => {
        switch (value) {
            case "AC":
                setInputValue("0")
                setPreviousValue(null)
                setOperator(null)
                break
            case "±":
                setInputValue(
                    (inputValue) =>
                        // inputValue.startsWith("-")
                        //     ? inputValue.slice(1)
                        //     : "-" + inputValue)
                        inputValue * -1
                )
                break
            case "%":
                setInputValue((inputValue) =>
                    (parseFloat(inputValue) / 100).toString()
                )
                break
            case ".":
                if (!inputValue.includes(".")) {
                    setInputValue((inputValue) => inputValue + ".")
                }
                break
            case "÷":
            case "x":
            case "-":
            case "+":
                setOperator(value)
                setPreviousValue(inputValue)
                setInputValue("0")
                break
            case "←":
                setInputValue((e) => (
                    e.length === 1
                        ? "0"
                        : 
                        e.slice(0, -1))
                    
                )
                break

            case "=":
                if (operator && previousValue && inputValue) {
                    let result
                    switch (operator) {
                        case "/":
                            result =
                                parseFloat(previousValue) /
                                parseFloat(inputValue)
                            break
                        case "x":
                            result =
                                parseFloat(previousValue) *
                                parseFloat(inputValue)
                            break
                        case "-":
                            result =
                                parseFloat(previousValue) -
                                parseFloat(inputValue)
                            result = result < 0 ? "ERROR" : result
                            break
                        case "+":
                            result =
                                parseFloat(previousValue) +
                                parseFloat(inputValue)
                            break
                        default:
                            break
                    }
                    const resultString = result.toString()
                    if (resultString.length > 9) {
                        setInputValue("ERROR")
                    } else {
                        setInputValue(resultString)
                    }
                    setPreviousValue(null)
                    setOperator(null)
                }
                break
            default:
                if (inputValue.length > 8) {
                    return
                }
                setInputValue(inputValue === "0" ? value : inputValue + value.toString())
                break
        }
    }

    return (
        <div className={style.Calculator}>
            <div className={style.Display} data-testid="display">
                {inputValue}
            </div>

            <div className={style.Top}>
                <Button
                    name="AC"
                    value={"AC"}
                    backgroundColor={"AC"}
                    setSelected={() => handleDisplay("AC")}
                />
                <Button
                    name="x"
                    value={"multiply"}
                    backgroundColor={"operator"}
                    setSelected={() => handleDisplay("x")}
                />
                <Button
                    name="÷"
                    value={"divide"}
                    backgroundColor={"operator"}
                    setSelected={() => handleDisplay("/")}
                />
                <Button
                    name="±"
                    value={"±"}
                    backgroundColor={"operator"}
                    setSelected={() => handleDisplay("±")}
                />
            </div>

            <div className={style.Bottom}>
                <div className={style.Numbers}>
                    {numbers.map((number) => (
                        <Button
                            name={number}
                            value={number}
                            setSelected={() => handleDisplay(number)}
                        />
                    ))}
                </div>
                <div className={style.Operations}>
                    <Button
                        name="+"
                        value={"add"}
                        backgroundColor={"operator"}
                        setSelected={() => handleDisplay("+")}
                    />

                    <Button
                        name="-"
                        value={"subtract"}
                        backgroundColor={"operator"}
                        setSelected={() => handleDisplay("-")}
                    />
                    <Button
                        name="="
                        value={"equal"}
                        backgroundColor={"equal"}
                        setSelected={() => handleDisplay("=")}
                    />
                </div>
            </div>
        </div>
    )
}
