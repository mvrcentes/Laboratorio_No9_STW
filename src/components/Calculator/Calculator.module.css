.Calculator {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    height: 800px;
    gap: 10px;
    padding: 15px;
    border-radius: 30px;
    background-image: linear-gradient(to top, #000000, #181718, #28282a, #383a3c, #484d4f);
}

.Display {
    display: flex;
    justify-content: flex-end;
    align-items: flex-end;
    font-size: 100px;
    font-weight: bold;
    height: 300px;
    width: 430px;
    border-radius: 15px;
    color: #a8aeae;
}

.Numbers {
    display: grid;
    grid-template-columns: repeat(3, 1fr);
    grid-template-rows: repeat(4, 1fr);
    grid-gap: 10px;
}

.Top {
    display: flex;
    flex-direction: row;
    gap: 10px;
}

.Bottom {
    display: flex;
    flex-direction: row;
    gap: 10px;
}

.Operations {
    display: flex;
    flex-direction: column;
    gap: 10px;
}