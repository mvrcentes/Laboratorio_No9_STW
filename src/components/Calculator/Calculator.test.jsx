import React from "react"
import { render, screen, fireEvent } from "@testing-library/react"

import { Calculator } from "./Calculator"

describe("Calculator", () => {
    it("should render correctly", () => {
        const { container } = render(<Calculator />)
        expect(container).toMatchSnapshot()

        screen.debug()
    })

    // Tests that clicking on "AC" button resets display and state.
    it("test_handle_display_resets_display_and_state", () => {
        render(<Calculator />)
        const acButton = screen.getByText("AC")
        acButton.click()

        // const display = screen.getByText("display")
        // expect(display.textContent).toEqual("0")
    })

    // Tests that clicking on numbers and operators updates display.
    it("test_handle_display_updates_display", () => {
        render(<Calculator />)

        fireEvent.click(screen.getByText("1"))
        expect(screen.getByTestId("display").textContent).toEqual("1")
        fireEvent.click(screen.queryAllByText("1")[1])
        expect(screen.getByTestId("display").textContent).toEqual("11")

        // fireEvent.click(screen.getByText("."))
        // expect(screen.getByTestId("display").textContent).toEqual("-.")

        // fireEvent.click(screen.getAllByText("0")[0])
        // expect(screen.getByTestId("display").textContent).toEqual("0")

        fireEvent.click(screen.getByText("←"))
        expect(screen.getByTestId("display").textContent).toEqual("1")

        // fireEvent.click(screen.getByText("÷"))
        // expect(screen.getByTestId("display").textContent).toEqual("/")
    })

    it("test_handle_display_updates_display_±", () => {
        render(<Calculator />)

        fireEvent.click(screen.getByText("1"))
        expect(screen.getByTestId("display").textContent).toEqual("1")

        fireEvent.click(screen.getByText("±"))
        expect(screen.getByTestId("display").textContent).toEqual("-1")
    })

    it("test_handle_display_updates_display_x", () => {
        render(<Calculator />)

        expect(screen.getByTestId("display").textContent).toEqual("0")

        fireEvent.click(screen.getByText("1"))
        expect(screen.getByTestId("display").textContent).toEqual("1")

        fireEvent.click(screen.getByText("x"))

        fireEvent.click(screen.getByText("2"))
        expect(screen.getByTestId("display").textContent).toEqual("2")

        fireEvent.click(screen.getByText("="))
        expect(screen.getByTestId("display").textContent).toEqual("2")
    })

    it("test_handle_display_updates_display_+", () => {
        render(<Calculator />)

        expect(screen.getByTestId("display").textContent).toEqual("0")

        fireEvent.click(screen.getByText("1"))
        expect(screen.getByTestId("display").textContent).toEqual("1")

        fireEvent.click(screen.getByText("+"))

        fireEvent.click(screen.getByText("2"))
        expect(screen.getByTestId("display").textContent).toEqual("2")

        fireEvent.click(screen.getByText("="))
        expect(screen.getByTestId("display").textContent).toEqual("3")
    })

    
})
