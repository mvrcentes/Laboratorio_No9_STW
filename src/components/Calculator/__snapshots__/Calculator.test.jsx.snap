// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Calculator > should render correctly 1`] = `
<div>
  <div
    class="_Calculator_818a78"
  >
    <div
      class="_Display_818a78"
      data-testid="display"
    >
      0
    </div>
    <div
      class="_Top_818a78"
    >
      <div
        class="_Button_2a1efe _Button__AC_2a1efe"
        value="AC"
      >
        AC
      </div>
      <div
        class="_Button_2a1efe _Button__operator_2a1efe"
        value="multiply"
      >
        x
      </div>
      <div
        class="_Button_2a1efe _Button__operator_2a1efe"
        value="divide"
      >
        ÷
      </div>
      <div
        class="_Button_2a1efe _Button__operator_2a1efe"
        value="±"
      >
        ±
      </div>
    </div>
    <div
      class="_Bottom_818a78"
    >
      <div
        class="_Numbers_818a78"
      >
        <div
          class="_Button_2a1efe _Button__undefined_2a1efe"
          value="7"
        >
          7
        </div>
        <div
          class="_Button_2a1efe _Button__undefined_2a1efe"
          value="8"
        >
          8
        </div>
        <div
          class="_Button_2a1efe _Button__undefined_2a1efe"
          value="9"
        >
          9
        </div>
        <div
          class="_Button_2a1efe _Button__undefined_2a1efe"
          value="4"
        >
          4
        </div>
        <div
          class="_Button_2a1efe _Button__undefined_2a1efe"
          value="5"
        >
          5
        </div>
        <div
          class="_Button_2a1efe _Button__undefined_2a1efe"
          value="6"
        >
          6
        </div>
        <div
          class="_Button_2a1efe _Button__undefined_2a1efe"
          value="1"
        >
          1
        </div>
        <div
          class="_Button_2a1efe _Button__undefined_2a1efe"
          value="2"
        >
          2
        </div>
        <div
          class="_Button_2a1efe _Button__undefined_2a1efe"
          value="3"
        >
          3
        </div>
        <div
          class="_Button_2a1efe _Button__undefined_2a1efe"
          value="."
        >
          .
        </div>
        <div
          class="_Button_2a1efe _Button__undefined_2a1efe"
          value="0"
        >
          0
        </div>
        <div
          class="_Button_2a1efe _Button__undefined_2a1efe"
          value="←"
        >
          ←
        </div>
      </div>
      <div
        class="_Operations_818a78"
      >
        <div
          class="_Button_2a1efe _Button__operator_2a1efe"
          value="add"
        >
          +
        </div>
        <div
          class="_Button_2a1efe _Button__operator_2a1efe"
          value="subtract"
        >
          -
        </div>
        <div
          class="_Button_2a1efe _Button__equal_2a1efe"
          value="equal"
        >
          =
        </div>
      </div>
    </div>
  </div>
</div>
`;
